hammer/hammer.go

```diff
@@ -0,0 +1,362 @@
+	for i := 0; i < *numReadersRandom; i++ {
+		h.randomReaders.Grow(ctx)
+	}
+	for i := 0; i < *numReadersFull; i++ {
+		h.fullReaders.Grow(ctx)
+	}
+	for i := 0; i < *numWriters; i++ {
+		h.writers.Grow(ctx)
+	}
+
+	// Set up logging for any errors
+	go func() {
+		for {
+			select {
+			case <-ctx.Done(): //context cancelled
+				return
+			case err := <-h.errChan:
+				klog.Warning(err)
+			}
+		}
+	}()
+
+	// Start the throttles
+	go h.readThrottle.Run(ctx)
+	go h.writeThrottle.Run(ctx)
+
+	go func() {
+		tick := time.NewTicker(1 * time.Second)
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case <-tick.C:
+				size := h.tracker.LatestConsistent.Size
+				_, _, _, err := h.tracker.Update(ctx)
+				if err != nil {
+					klog.Warning(err)
+					inconsistentErr := client.ErrInconsistency{}
+					if errors.As(err, &inconsistentErr) {
+						klog.Fatalf("Last Good Checkpoint:\n%s\n\nFirst Bad Checkpoint:\n%s\n\n%v", string(inconsistentErr.SmallerRaw), string(inconsistentErr.LargerRaw), inconsistentErr)
+					}
+				}
+				newSize := h.tracker.LatestConsistent.Size
+				if newSize > size {
+					klog.V(1).Infof("Updated checkpoint from %d to %d", size, newSize)
+				}
+			}
+		}
+	}()
+}
+
+func genLeaf(n uint64, minLeafSize int) []byte {
+	// Make a slice with half the number of requested bytes since we'll
+	// hex-encode them below which gets us back up to the full amount.
+	filler := make([]byte, minLeafSize/2)
+	_, _ = crand.Read(filler)
+	return []byte(fmt.Sprintf("%x %d", filler, n))
+}
+
+func newLeafGenerator(n uint64, minLeafSize int) func() []byte {
+	const dupChance = 0.1
+	nextLeaf := genLeaf(n, minLeafSize)
+	return func() []byte {
+		if rand.Float64() <= dupChance {
+			// This one will actually be unique, but the next iteration will
+			// duplicate it. In future, this duplication could be randomly
+			// selected to include really old leaves too, to test long-term
+			// deduplication in the log (if it supports  that).
+			return nextLeaf
+		}
+
+		n++
+		r := nextLeaf
+		nextLeaf = genLeaf(n, minLeafSize)
+		return r
+	}
+}
+
+func NewThrottle(opsPerSecond int) *Throttle {
+	return &Throttle{
+		opsPerSecond: opsPerSecond,
+		tokenChan:    make(chan bool, opsPerSecond),
+	}
+}
+
+type Throttle struct {
+	opsPerSecond int
+	tokenChan    chan bool
+
+	oversupply int
+}
+
+func (t *Throttle) Increase() {
+	tokenCount := t.opsPerSecond
+	delta := float64(tokenCount) * 0.1
+	if delta < 1 {
+		delta = 1
+	}
+	t.opsPerSecond = tokenCount + int(delta)
+}
+
+func (t *Throttle) Decrease() {
+	tokenCount := t.opsPerSecond
+	if tokenCount <= 1 {
+		return
+	}
+	delta := float64(tokenCount) * 0.1
+	if delta < 1 {
+		delta = 1
+	}
+	t.opsPerSecond = tokenCount - int(delta)
+}
+
+func (t *Throttle) Run(ctx context.Context) {
+	ticker := time.NewTicker(1 * time.Second)
+	for {
+		select {
+		case <-ctx.Done(): //context cancelled
+			return
+		case <-ticker.C:
+			tokenCount := t.opsPerSecond
+			timeout := time.After(1 * time.Second)
```

Comment:
I think this 1*Second would need to change if the `ticker` changed too - worth factoring the duration out to an `interval` var at the top of the func?

Reply:
Done.